src/thor/route_action.cc

```diff
@@ -489,50 +513,87 @@ void thor_worker_t::path_arrive_by(Api& api, const std::string& costing) {
     // if we just made a leg that means we are done recording which algorithms were used
     if (path.empty())
       algorithms.clear();
-  }
 
+    return true;
+  };
+
+  auto correlated = options.locations();
+  bool allow_retry = true;
+
+  // For each pair of locations
+  auto origin = ++correlated.rbegin();
+  while (origin != correlated.rend()) {
+    auto destination = std::prev(origin);
+    if (!route_two_points(origin, destination)) {
+      // if routing failed because an intermediate waypoint was snapped to the low reachability road
+      // (such road lies in a small connectivity component that is not connected to other locations)
+      // we should leave only high reachability candidates and try to route again
+      if (allow_retry && is_through_point(*destination) &&
+          !is_high_reachable(*destination, destination->path_edges(0))) {
```

Comment:
the reason why `0` works here is only because its a `through` which means we've culled all other edges but the matching ones? what about on the first leg of the route where we havent dont any culling yet, why does `0` always work here?

Reply:
hm, good question. I though that the first (last) point can't be a `through` type

Reply:
youre right it cant!

Reply:
should we handle it somehow ?